### DISCUSSION TITLE
Generate setters for used Mappers?

### DISCUSSION THREAD

**Discussion by obfischer** [mapstruct/mapstruct#3497] [Ideas]
If I write a mapper with component model `SPRING`, which uses other mappers, like this:

```java
@Mapper(componentModel = MappingConstants.ComponentModel.CDI, uses = ReferenceMapper.class )
public interface CarMapper {
    Car carDtoToCar(CarDto carDto);
}
```

the used mapper `ReferenceMapper` is injected via field or constructor injection. This works fine for production code.

But in tests, it would be great to have also a setter for every used mapper, so that I can setup a test as unit test outside of a Spring context. Currently I have to annotate with `@SpringBootTest`, even if I don't need the Spring context.

So, would it be possible to generate setters for each used mapper? IMHO it would ease writing simple and fast tests.

**Comment by filiphr** [comment]
When you are using `InjectionStrategy#CONSTRUCTOR` you can create tests outside of a Spring context fast.

In any case in 1.6 (also part of 1.6.0.Beta1) there is `InjectionStrategy#SETTER` that can be used to instruct MapStruct to create setters instead of field injection.

Will something like this work for you?

**Reply by obfischer** [reply]
Yes, it will. Thank you

**Reply by obfischer** [reply]
Hi @filiphr, 

I am currently using 1.6.0.Beta1 and it is not working as I need it. Currently I have to use static mocking via Mockito to be able to configure used mappers.

Would you have a look at it, if I would provide an example project? The actual project is closed source.

**Reply by filiphr** [reply]
We'll have a look at an example project @obfischer. Please share it with us and we will see what we can improve

**Reply by obfischer** [reply]
Hi @filiphr, while preparing an example, I was able to figure out how to configure all the mappers so, that I can assign a mapper via a setter.

**Reply by obfischer** [reply]
But only one question more: If you look at my example project at https://github.com/obfischer/playground-mapstruct/, you can see a main mapper and a used mapper. 

The used mapper (`B2YMapper`) is not automatically apply for a mapping from `B` to `Y`, so that I have to add `@Mapping(target = "y", source = "b")`. Is it already possible to apply all used mappers automatically?

**Reply by thunderhook** [reply]
> The used mapper (`B2YMapper`) is not automatically apply for a mapping from `B` to `Y`, so that I have to add `@Mapping(target = "y", source = "b")`. Is it already possible to apply all used mappers automatically?

Two things about this with your example modified:

```java
// public modifiers for brewity
class A {
  public B b;
  public B anotherB;
  public B field1;
  public B field2;
  public B field3;
}
class Z {
  public Y y;
  public Y field1;
  public Y field2;
  public Y field3;
}
```

1. When you define a mapping method from `B` to `Y`, everywhere where a mapping like this **can** be used, **will** be used.
2. But: Mapstruct implicilty (=automatically) maps properties that have the **same name**. So in the given example only `field1, field2, field3` get mapped but `b -> y` will not. For various reasons, some of them coming to my mind:
   * What should be the source of `y`? `b` or `anotherB`?
   * What if there is `C` that could also be mapped into `Y` - should this be considered? Which one has a higher priority?
   * This would also very non-transparent for the user what is getting mapped and why (especially for complex objects)

So if you have the possibility/luxury of naming your objects yourself, and you want to make it easier for MapStruct, then use the same property names (if applicable).

-----

So in your example, the solution would be to have the same names in source and target object - looks really weird in your inital example, but for the sake of completeness, this would automatically map:

```java
class A {
  public B b;
}
class Z {
  public Y b; // <--
}
```

**Comment by obfischer** [comment]
Thanks to all of you. I think my mistake was that I focused on type compatibility. But for MapStruct the name of the property is the main focus. 

So thank you very much for MapStruct
